styx-common/src/main/java/com/spotify/styx/util/GoogleIdTokenValidator.java

```diff
@@ -0,0 +1,218 @@
+                         Iam iam,
+                         Set<String> domainWhitelist) {
+    this.googleIdTokenVerifier =
+        Objects.requireNonNull(googleIdTokenVerifier, "googleIdTokenVerifier");
+    this.cloudResourceManager =
+        Objects.requireNonNull(cloudResourceManager, "cloudResourceManager");
+    this.iam = Objects.requireNonNull(iam, "iam");
+    this.domainWhitelist = Objects.requireNonNull(domainWhitelist, "domainWhitelist");
+  }
+
+  public void cacheProjects() throws IOException {
+    final CloudResourceManager.Projects.List request = cloudResourceManager.projects().list();
+
+    ListProjectsResponse response;
+    do {
+      response = request.execute();
+      if (response.getProjects() == null) {
+        continue;
+      }
+      for (Project project : response.getProjects()) {
+        projectCache.add(project.getProjectId());
+      }
+      request.setPageToken(response.getNextPageToken());
+    } while (response.getNextPageToken() != null);
+
+    logger.info("project cache loaded");
+  }
+
+  public GoogleIdToken validate(String token) {
+    final GoogleIdToken googleIdToken;
+    try {
+      googleIdToken = verifyIdToken(token);
+    } catch (IOException e) {
+      logger.warn("failed to verify token");
+      return null;
+    }
+
+    if (googleIdToken == null) {
+      return null;
+    }
+
+    final String email = googleIdToken.getPayload().getEmail();
+
+    final Matcher matcher = EMAIL_PATTERN.matcher(email);
+    if (matcher.matches()) {
+      final String domain = matcher.group(1);
+      if (domainWhitelist.contains(domain)) {
+        logger.debug("domain {} in whitelist", domain);
+        return googleIdToken;
+      }
+    } else {
+      logger.warn("invalid email address {}", email);
+      return null;
+    }
+
+    if (validatedEmailCache.getIfPresent(email) != null) {
+      return googleIdToken;
+    }
+
+    final String projectId;
+    try {
+      projectId = checkProject(email);
+    } catch (IOException e) {
+      logger.info("cannot validate {}", email);
+      return null;
+    }
+
+    if (projectId != null) {
+      validatedEmailCache.put(email, projectId);
+      return googleIdToken;
+    }
+
+    return null;
+  }
+
+  @VisibleForTesting
+  void clearProjectCache() {
+    projectCache.clear();
+  }
+
+  private GoogleIdToken verifyIdToken(String token) throws IOException {
+    try {
+      return googleIdTokenVerifier.verify(token);
+    } catch (GeneralSecurityException e) {
+      return null;
+    }
+  }
+
+  private String checkProject(String email) throws IOException {
+    String projectId = null;
+
+    final Matcher matcher = SERVICE_ACCOUNT_PATTERN.matcher(email);
+    if (matcher.matches()) {
+      projectId = matcher.group(1);
+    }
+
+    if (projectId == null) {
+      // no projectId, could be GCE default
+      logger.debug("{} doesn't contain project id, try getting its project", email);
+      projectId = getProjectIdOfServiceAccount(email);
```

Comment:
I think in many cases, we will receive a normal service account to validate. It feels beneficial to avoid making that API call. The tradeoff is of course less secure.